src/server/controllers/authController.js

```diff
@@ -26,33 +31,64 @@ const refreshTokenExpirySecs = refreshTokenExpiryMillis / 1000;
 async function handleLogin(req, res) {
     if (!verifyBodyHasLoginFormData(req)) return; // If false, it will have already sent a response.
 
+    
     let { username, password } = req.body;
     const usernameLowercase = username.toLowerCase();
     const usernameCaseSensitive = getUsernameCaseSensitive(usernameLowercase); // False if the member doesn't exist
     const hashedPassword = getHashedPassword(usernameLowercase);
 
     if (!usernameCaseSensitive || !hashedPassword) return res.status(401).json({ 'message': 'Username or password is incorrect'}); // Unauthorized, username not found
+    
+    clientIP = getClientIP(req);
+    if(!(clientIP in loginAttemptData)) {
+        loginAttemptData[clientIP] = {};
+    }
+
+    if(!(usernameLowercase in loginAttemptData[clientIP])) {
+        loginAttemptData[clientIP][usernameLowercase] = { attempts: 1, cooldownTimeSec: 0 };
```

Comment:
Hmm.. a 2D array I think is a little too complex for this thing. I suggest doing something like this:

```js
const browserAgent = `${usernameLowercase}; ${clientIP}`;
if(!(browserAgent in loginAttemptData)) {
        loginAttemptData[browserAgent] = { attempts: 1, cooldownTimeSec: 0 };
    }
```